File: packages/pigeon/platform_tests/alternate_language_test_plugin/android/src/test/java/com/example/alternate_language_test_plugin/ListTest.java
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.example.alternate_language_test_plugin;

import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

import com.example.alternate_language_test_plugin.CoreTests.FlutterSmallApi;
import com.example.alternate_language_test_plugin.CoreTests.TestMessage;
import io.flutter.plugin.common.BinaryMessenger;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import org.junit.Test;

public class ListTest {
  @Test
  public void listInList() {
    TestMessage top = new TestMessage();
    TestMessage inside = new TestMessage();
    inside.setTestList(Arrays.asList(1, 2, 3));
    top.setTestList(Arrays.asList(inside));
    BinaryMessenger binaryMessenger = mock(BinaryMessenger.class);
    doAnswer(
            invocation -> {
              ByteBuffer message = invocation.getArgument(1);
              BinaryMessenger.BinaryReply reply = invocation.getArgument(2);
              message.position(0);
              ArrayList args = (ArrayList) FlutterSmallApi.getCodec().decodeMessage(message);
              ByteBuffer replyData = FlutterSmallApi.getCodec().encodeMessage(args.get(0));
              replyData.position(0);
              reply.reply(replyData);
              return null;
            })
        .when(binaryMessenger)
        .send(anyString(), any(), any());
    FlutterSmallApi api = new FlutterSmallApi(binaryMessenger);
    boolean[] didCall = {false};
    api.echoWrappedList(
        top,
        (result) -> {
          didCall[0] = true;
          assertEquals(result.getTestList().size(), 1);
          assertTrue(result.getTestList().get(0) instanceof TestMessage);
          TestMessage readInside = (TestMessage) result.getTestList().get(0);
          assertEquals(readInside.getTestList().size(), 3);
        });
    assertTrue(didCall[0]);
  }
}

File: packages/pigeon/platform_tests/test_plugin/android/src/test/kotlin/com/example/test_plugin/ListTest.kt
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.example.test_plugin

import io.flutter.plugin.common.BinaryMessenger
import io.mockk.every
import io.mockk.mockk
import java.nio.ByteBuffer
import java.util.ArrayList
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class ListTest {

  @Test
  fun testListInList() {
    val binaryMessenger = mockk<BinaryMessenger>()
    val api = FlutterSmallApi(binaryMessenger)

    val inside = TestMessage(listOf(1, 2, 3))
    val input = TestMessage(listOf(inside))

    every { binaryMessenger.send(any(), any(), any()) } answers
        {
          val codec = FlutterSmallApi.codec
          val message = arg<ByteBuffer>(1)
          val reply = arg<BinaryMessenger.BinaryReply>(2)
          message.position(0)
          val args = codec.decodeMessage(message) as ArrayList<*>
          val replyData = codec.encodeMessage(args)
          replyData?.position(0)
          reply.reply(replyData)
        }

    var didCall = false
    api.echoWrappedList(input) {
      didCall = true
      assertEquals(input, it.getOrNull())
    }

    assertTrue(didCall)
  }
}
